### PR TITLE
[checkpoint] Open Source

### PR DESCRIPTION
**Pull request by MingjiHan99**
### In this PR, we open source our `vescale.checkpoint`, Yo. ~

`vescale.checkpoint` is a distributed LLM checkpointing system.  

`vescale.checkpoint` offers simple and straightforward APIs,
enabling users to load and save distributed model (`DModule`) and optimizer (`DistributedOptimizer`) seamlessly, 
abstracting away the complexities of underlying details such as process rank and device mesh.  

`vescale.checkpoint`supports load-time checkpoint resharding when varying the degrees of data, tensor, or pipeline (TODO) parallelism 
for both veScale distributed model (`DModule`) and optimizer (`DistributedOptimizer`).  

`vescale.checkpoint` incorporates [fast checkpointing](https://arxiv.org/abs/2402.15627) and various I/O optimization techinques, 
enhancing I/O efficiency during large language model training.  

`vescale.checkpoint` will be a part of OmniStore project, a new open source project coming soon.  

### Credit to veScale Checkpoint Team
This endeavor would not have been possible without the contribution of veScale Checkpoint team which includes but not limited to:
@shanesyy-1992 @MingjiHan99 @AHEADer @raywan-110 @michael4RD @lazychao @leochen-ai 

Also thanks to the great guidance and leadership of: [@pengyanghua](https://github.com/pengyanghua) [@eric-haibin-lin](https://github.com/eric-haibin-lin) [@liwenchangbdbz](https://github.com/liwenchangbdbz) [@Meteorix](https://github.com/Meteorix)

### Credit to veScale Team

We would like to sincerely acknowledge the assistance of and collaboration with the veScale team which inlcudes but not limited to:
[@leonardo0lyj](https://github.com/leonardo0lyj) @JsBlueCat @MackZackA @Vremold @jc-bytedance @lichen225

### Credit to PyTorch Distributed Checkpoint (DCP) Team
We would like to sincerely acknowledge the assistance of and collaboration 
with the [PyTorch Distributed Checkpoint (DCP) team](https://github.com/pytorch/pytorch/tree/main/torch/distributed/checkpoint) 
which includes but not limited to:
@wz337 @kumpera @fegin @LucasLLC

### CODE REVIEW

File: python/vescale/checkpoint/api/base_checkpointer.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    def save(cls, path: str, checkpoint_state: CheckpointState):
+        """
+        A Method for saving checkpoint
+        Args:
+            path: Defines the storage path for checkpoint.
+            checkpoint_state: A dictionary contains key-value pairs for model, optimizer and dataloader(TODO).
+                              - Model: Identified by 'model' key, value should be a model instance.
+                              - Optimizer: Identified by 'optimizer' key, value should be an optimizer instance.
+        """
+        pass
```
Comment (JsBlueCat) [inline python/vescale/checkpoint/api/base_checkpointer.py:36]: should we raise unimplement error?

File: python/vescale/checkpoint/api/base_checkpointer.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    def load(cls, path: str, checkpoint_state: CheckpointState):
+        """
+        A Method for loading checkpoint
+        Args:
+            path: Defines the storage path for checkpoint.
+            checkpoint_state: A dictionary contains key-value pairs for model, optimizer and dataloader(TODO).
+                              - Model: Identified by 'model' key, value should be a model instance.
+                              - Optimizer: Identified by 'optimizer' key, value should be an optimizer instance.
+        """
+        pass
```
Comment (JsBlueCat) [inline python/vescale/checkpoint/api/base_checkpointer.py:47]: should we raise unimplement error?

**Review by JsBlueCat — COMMENTED** [review]
I have some questions abort protobuf, does we need that codegen file to be push?

File: python/vescale/checkpoint/api/base_checkpointer.py (hunk 1 above)
Comment (MingjiHan99) [reply python/vescale/checkpoint/api/base_checkpointer.py:36]: Yes

File: python/vescale/checkpoint/api/base_checkpointer.py (hunk 2 above)
Comment (MingjiHan99) [reply python/vescale/checkpoint/api/base_checkpointer.py:47]: Yes

**Comment by MingjiHan99** [comment]
> I have some questions abort protobuf, does we need that codegen file to be push?

Based on our discussion, we reserve the protobuf files for now.

**Review by leonardo0lyj — APPROVED** [review]

**Review by jc-bytedance — APPROVED** [review]

**Comment by shanesyy-1992** [comment]
Could you help make some clean with the fast checkpoint code. There seems to be some code that hasn't been used.

**Comment by MingjiHan99** [comment]
> Could you help make some clean with the fast checkpoint code. There seems to be some code that hasn't been used.

Sure. I will remove `DistributedTorchLoader` and `RemappingTorchLoader`.

**Review by shanesyy-1992 — APPROVED** [review]

**Pull request merged**